Custom derive fails on structs with type parameters

## How to reproduce

```rust
#[derive(Parser)]
pub struct Parser<'a, T>;
```

generates the compile-time errors `error[E0106]: missing lifetime specifier` and `error[E0243]: wrong number of type arguments`, but confusingly points to the identifier `Parser` as the source of the errors.

## Expected behavior

The custom derivation would work on any struct, regardless of type parameters, or else fail with a descriptive error message.

--- a/derive/src/generator.rs
+++ b/derive/src/generator.rs
@@ -10,11 +10,11 @@
 use std::collections::HashMap;
 
 use quote::Tokens;
-use syn::Ident;
+use syn::{Generics, Ident};
 
 use pest_meta::ast::*;
 
-pub fn generate(name: Ident, rules: Vec<Rule>, defaults: Vec<&str>) -> Tokens {
+pub fn generate(name: Ident, generics: &Generics, rules: Vec<Rule>, defaults: Vec<&str>) -> Tokens {
     let mut predefined = HashMap::new();
     predefined.insert(
         "any",
@@ -87,8 +87,10 @@ pub fn generate(name: Ident, rules: Vec<Rule>, defaults: Vec<&str>) -> Tokens {
             .map(|name| predefined.get(name).unwrap().clone())
     );
 
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
     let parser_impl = quote! {
-        impl ::pest::Parser<Rule> for #name {
+        impl #impl_generics ::pest::Parser<Rule> for #name #ty_generics #where_clause {
             fn parse<'i>(
                 rule: Rule,
                 input: &'i str
@@ -850,6 +852,7 @@ mod tests {
     #[test]
     fn generate_complete() {
         let name = Ident::from("MyParser");
+        let generics = Generics::default();
         let rules = vec![
             Rule {
                 name: "a".to_owned(),
@@ -860,7 +863,7 @@ mod tests {
         let defaults = vec!["any"];
 
         assert_eq!(
-            generate(name, rules, defaults),
+            generate(name, &generics, rules, defaults),
             quote! {
                 #[allow(dead_code, non_camel_case_types)]
                 #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]

--- a/derive/src/lib.rs
+++ b/derive/src/lib.rs
@@ -256,7 +256,7 @@ use std::io::{self, Read};
 use std::path::Path;
 
 use proc_macro::TokenStream;
-use syn::{Attribute, DeriveInput, Ident, Lit, Meta};
+use syn::{Attribute, DeriveInput, Generics, Ident, Lit, Meta};
 
 mod generator;
 mod optimizer;
@@ -267,7 +267,7 @@ use pest_meta::parser::{self, Rule};
 #[proc_macro_derive(Parser, attributes(grammar))]
 pub fn derive_parser(input: TokenStream) -> TokenStream {
     let ast: DeriveInput = syn::parse(input).unwrap();
-    let (name, path) = parse_derive(ast);
+    let (name, generics, path) = parse_derive(ast);
 
     let root = env::var("CARGO_MANIFEST_DIR").unwrap_or(".".into());
     let path = Path::new(&root).join("src/").join(&path);
@@ -318,7 +318,7 @@ pub fn derive_parser(input: TokenStream) -> TokenStream {
     let defaults = unwrap_or_report(validator::validate_pairs(pairs.clone()));
     let ast = unwrap_or_report(parser::consume_rules(pairs));
     let optimized = optimizer::optimize(ast);
-    let generated = generator::generate(name, optimized, defaults);
+    let generated = generator::generate(name, &generics, optimized, defaults);
 
     generated.into()
 }
@@ -330,8 +330,9 @@ fn read_file<P: AsRef<Path>>(path: P) -> io::Result<String> {
     Ok(string)
 }
 
-fn parse_derive(ast: DeriveInput) -> (Ident, String) {
+fn parse_derive(ast: DeriveInput) -> (Ident, Generics, String) {
     let name = ast.ident;
+    let generics = ast.generics;
 
     let grammar: Vec<&Attribute> = ast.attrs
         .iter()
@@ -347,7 +348,7 @@ fn parse_derive(ast: DeriveInput) -> (Ident, String) {
         _ => panic!("only 1 grammar file can be provided")
     };
 
-    (name, filename)
+    (name, generics, filename)
 }
 
 fn get_filename(attr: &Attribute) -> String {
@@ -373,7 +374,7 @@ mod tests {
             pub struct MyParser<'a, T>;
         ";
         let ast = syn::parse_str(definition).unwrap();
-        let (_, filename) = parse_derive(ast);
+        let (_, _, filename) = parse_derive(ast);
         assert_eq!(filename, "myfile.pest");
     }
 

